ir/memory.cpp

```diff
@@ -248,16 +248,37 @@ pair<expr, vector<expr>> Memory::mkInput(const char *name) {
   return { Pointer(*this, offset, local_bid, bid).release(), { var } };
 }
 
-expr Memory::alloc(const expr &size, unsigned align, bool heap) {
-  Pointer p(*this, ++last_bid, true);
+expr Memory::alloc(const expr &size, unsigned align, BlockKind blockKind,
+                   optional<unsigned> bidopt, unsigned *bid_out) {
+  unsigned bid = bidopt ? *bidopt : last_bid;
+  if (last_bid <= bid) {
+    last_bid = bid + 1;
+  }
+  if (bid_out)
+    *bid_out = bid;
+
+  // It is allocated as a local block if blockKind is heap or stack.
+  bool is_local = blockKind != GLOBAL;
+
+  Pointer p(*this, bid, is_local);
   state->addPre(p.is_aligned(align));
 
   expr size_zextOrTrunced = size.zextOrTrunc(bits_size_t);
   state->addPre(p.block_size() == size_zextOrTrunced);
+  // TODO: If its address space is not 0, its address can be 0.
+  state->addPre(p.get_address() != 0);
```

Comment:
wait, I thought the conclusion is that malloc can return null?

Reply:
I believe this can be addressed by making Malloc::toSMT return `mkIf(random_bool, Memory::alloc(), nullptr)`. I will make a separate PR for this so discussion for its implementation can be done there.